### ISSUE TITLE
type assertions

### ISSUE THREAD

**Issue by thecotne** [JSMonk/hegel#190]
i am doing something like this in flow

```js
type NotificationAction =
  | 'ScheduleCreated'
  | 'ScheduleStarttimeUpdated'
  | 'ScheduleCancelled'
  | 'DiagnosisPublished'
  | 'PatientCaseFinished'
  | 'MydocGreeted'

type NotificationBase = {|
  +id: string,
  +created_at: string
|}

type Notification = {|
  ...NotificationBase,
  +action: NotificationAction,
  +entity_id: string,
|}

export type NotificationExpanded =
  | {|
      ...NotificationBase,
      +action: 'ScheduleCreated' | 'ScheduleStarttimeUpdated' | 'ScheduleCancelled',
      +schedule: {...}
    |}
  | {|
      ...NotificationBase,
      +action: 'PatientCaseFinished',
      +patientCase: {...}
    |}
  | {|
      ...NotificationBase,
      +action: 'MydocGreeted' | 'DiagnosisPublished'
    |}

// -- EXHAUSTIVE CHECKING --

declare var _asrt01: $PropertyType<NotificationExpanded, 'action'>
/*:: (_asrt01: NotificationAction) */

declare var _asrt02: NotificationAction
/*:: (_asrt02: $PropertyType<NotificationExpanded, 'action'> ) */

// -- END OF EXHAUSTIVE CHECKING --
```

[try](https://flow.org/try/#0C4TwDgpgBAcg9sAlgM0QYwIZLgOwIJrY5QC8AUFFAD5QDkAymgBYQAmArgDYQDCAThCxtaFanUYsO3esAx9gSALYQAqmFZDWIyjQbM2XXhhxoInbltG6AIogwBzHHADOiZwAV2AI05vJ2sVp3LEQIHGAeDGcIADFEHD9hKzoAWRBWODQAcQEIYCSyUEhYBBR0ENwAISjoEigAbypRAGpEVgAuKGdgPnj7ABoWtAFNAH0sTu7enHsyKgBfMkLwaHgkVEwiUgamygA6A7WyzcQqmsHKZoxCU5xOo42K-BvcC6hmsKRQUbbJnr7BgslhAAB5gODyKBFValR5EACiYOMrDYpGSjVElH2h1h5SI1Wibyx72uRE6ekkhn4gnyWkCEgM0lk8iUqnUmlo9P0UiMJjMFloRKxzWc3MMnXqBz2i2JQJ0O0xWKlDzxtwJECFl1Jt3JwSQn0i0TiCWc-k17zAIQNNQlUplWLlYgxxOxexVJzOhMVWped1S6UyOQgeWEgVsDicrg83l8pqSssWZAA9EmoABaNNQeEADQAEngVPQACoASQAavCoDxc-CeABpEswLLptNLFFoThyaAANzkUHGznkAAYAIydAAk7j4cEg8hARZWAB53U9EZacCjWP06NrcLQAHzJgBU7U6AAoB8OxyV1qrcAQiABKKBHpNtiAdrtQXt8ftRYcAEz3LiHrPEQx6nlAF7-sAQ5AVAk7TrOoALpAy4gauSIbmw260LuOAHlAz6vksKYtlmMDWFAADyMRZnmBbFuWlbVrWDZNi2ZBAA)

i am checking that object union has all actions from `NotificationAction` and nothing extra

in flow this is very hacky but in hegel i think it's impossible to achieve same result because i can't `declare var` and i can't use type casting (type casting can be replaced with variable declaration but then i have reference of undefined value in code)

**Comment by JSMonk** [comment]
Good point. I need to think about it, especially about the syntax. Thank you for your contribution ^_^.
